projects/graph/src/graph.py

```diff
@@ -2,52 +2,270 @@
-        explored = []
-        # Keep track of nodes to be checked
-        queue = [start]
-
-        # Keep looping until there are nodes still to be checked
-        while queue:
-            # Pop shallowest node (first node) from queue
-            node = queue.pop(0)
-            if node not in explored:
-                # Add node to list of checked nodes
-                explored.append(node)
-                neighbours = self.verticies[node]
-
-                # Add neighbors of node to queue
-                for neighbour in neighbours:
-                    queue.append(neighbour)
-        return explored
-
-
-graph = Graph()
-graph.add_vertex('0')
-graph.add_vertex('1')
-graph.add_vertex('2')
-graph.add_vertex('3')
-graph.add_edge('0', '1')
-graph.add_edge('0', '3')
-print(graph.verticies)
-
-print(graph.bfs('1'))
+        self.vertices = {}
+
+    def add_vertex(self, vertex_id):
+        self.vertices[vertex_id] = Vertex(vertex_id)
+
+    def add_edge(self, v1, v2):
+        if v1 in self.vertices and v2 in self.vertices:
+            self.vertices[v1].edges.add(v2)
+            self.vertices[v2].edges.add(v1)
+        else:
+            raise IndexError("That vertex does not exist")
+
+    def add_directed_edge(self, v1, v2):
+        if v1 in self.vertices and v2 in self.vertices:
+            self.vertices[v1].edges.add(v2)
+        else:
+            raise IndexError("That vertex does not exist")
+
+    def bft(self, start):
+        # Create empty Queue
+        q = Queue()
+        # Create an empty visited list
+        visited = set()
+        # Add the start node to the queue
+        q.enqueue(start)
+        # While the Queue is not empty...
+        while q.size() > 0:
+            # Remove the first node from the Queue
+            node = q.dequeue()
+            # If it hasn't been visited
+            if node not in visited:
+                # Mark it as visited
+                print(node)
+                visited.add(node)
+                # then put all it's children in the queue
+                for child in self.vertices[node].edges:
+                    q.enqueue(child)
+
+    def dft(self, start):
+        # Create empty Stack
+        s = Stack()
+        # Create an empty visited list
+        visited = set()
+        # Add the start node to the stack
+        s.push(start)
+        # While the Stack is not empty...
+        while s.size() > 0:
+            # Remove the first node from the Queue
+            node = s.pop()
+            # If it hasn't been visited
+            if node not in visited:
+                # Mark it as visited
+                print(node)
+                visited.add(node)
+                # then put all it's children in the stack
+                for child in self.vertices[node].edges:
+                    s.push(child)
+
+    # I could probably simplify this more
+    def dft_r(self, node, visited=None, s=None):
+        if visited is None:
+            visited = set()
```

Comment:
It's a small thing, but you could save yourself two lines by making the default value of visited an empty list